docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst

```diff
@@ -0,0 +1,380 @@
+
+-  The class ``RemoteFlowSource`` (defined in module ``semmle.python.dataflow.new.RemoteFlowSources``) represents data flow from remote network inputs. This is useful for finding security problems in networked services.
+-  The library ``Concepts`` (defined in module ``semmle.python.Concepts``) contain several subclasses of ``DataFlow::Node`` that are security relevant, such as ``FileSystemAccess`` and ``SqlExecution``.
+-  The module ``Attributes`` (defined in module ``semmle.python.dataflow.new.internal.Attributes``) defines ``AttrRead`` and ``AttrWrite`` which handle both ordinary and dynamic attribute access.
+
+For global flow, it is also useful to restrict sources to instances of ``LocalSourceNode``. The predefined sources generally do that.
+
+Class hierarchy
+~~~~~~~~~~~~~~~
+
+-  ``DataFlow::Configuration`` - base class for custom global data flow analysis.
+-  ``DataFlow::Node`` - an element behaving as a data flow node.
+
+    -  ``DataFlow::CfgNode`` - a control-flow node behaving as a data flow node.
+
+        -  ``DataFlow::ExprNode`` - an expression behaving as a data flow node.
+        -  ``DataFlow::ParameterNode`` - a parameter data flow node representing the value of a parameter at function entry.
+        -  ``DataFlow::CallCfgNode`` - a control-flow node for a function or method call behaving as a data flow node.
+
+    -  ``RemoteFlowSource`` - data flow from network/remote input.
+    -  ``Attributes::AttrRead`` - flow out of an attribute.
+    -  ``Attributes::AttrWrite`` - flow into an attribute.
+    -  ``Concepts::SystemCommandExecution`` - a data-flow node that executes an operating system command, for instance by spawning a new process.
+    -  ``Concepts::FileSystemAccess`` - a data flow node that performs a file system access, including reading and writing data, creating and deleting files and folders, checking and updating permissions, and so on.
+    -  ``Concepts::Path::PathNormalization`` - a data-flow node that performs path normalization. This is often needed in order to safely access paths.
+    -  ``Concepts::Decoding`` - a data-flow node that decodes data from a binary or textual format. A decoding (automatically) preserves taint from input to output. However, it can also be a problem in itself, for example if it allows code execution or could result in denial-of-service.
+    -  ``Concepts::Encoding`` - a data-flow node that encodes data to a binary or textual format. An encoding (automatically) preserves taint from input to output.
+    -  ``Concepts::CodeExecution`` - a data-flow node that dynamically executes Python code.
+    -  ``Concepts::SqlExecution`` - a data-flow node that executes SQL statements.
+    -  ``Concepts::HTTP::Server::RouteSetup`` - a data-flow node that sets up a route on a server.
+    -  ``Concepts::HTTP::Server::HttpResponse`` - a data-flow node that creates a HTTP response on a server.
```

Comment:
Continuing the discussion from https://github.com/github/codeql/pull/5357#discussion_r595090529, I think with the addition of this:

> The library ``Concepts`` (defined in module ``semmle.python.Concepts``) contain several subclasses of ``DataFlow::Node`` that are security relevant, such as ``FileSystemAccess`` and ``SqlExecution``.

we don't need to talk a great deal more about Concepts. Specifically, I think the list becomes too bloated with all those concepts, so I would recommend we don't include them. (I'm probably feeling 5/10 about this)

```suggestion
```

Reply:
So where do you propose this list be available? Or do you mean that no-one should maintain it and users should simply trawl `Concepts.qll` or rely on auto-complete?

Reply:
discussed this in person, @yoff is feeling much stronger about this, so I'll bend to his will :+1: